### PR TITLE
[WIP] MINT-1743 Rewrite bitcoin dao to redux

### PR DESCRIPTION
**Pull request by ex1st0r**

### CODE REVIEW

File: packages/core/redux/bitcoin/thunks.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+      reject: rejectTransaction,
+    },
+  }))
+}
+
+const acceptTransaction = (entry) => async (dispatch, getState) => {
+  dispatch(BitcoinActions.acceptTransaction(entry))
+
+  const state = getState()
+  const signer = getSigner(state)
```
Comment (Mikefluff) [inline packages/core/redux/bitcoin/thunks.js]: we should getBtcSigner here and use it to sign btc transactions

File: packages/core/redux/wallets/actions.js [hunk 2]
```diff
@@ -209,13 +210,14 @@ const updateAllowance = (allowance) => (dispatch, getState) => {
...
 
-export const mainTransfer = (wallet: WalletModel, token: TokenModel, amount: Amount, recipient: string, feeMultiplier: Number = 1) => async (dispatch) => {
+export const mainTransfer = (wallet: WalletModel, token: TokenModel, amount: Amount, recipient: string, feeMultiplier: Number = 1, advancedParams = null) => async (dispatch) => {
   try {
     const tokenDAO = tokenService.getDAO(token.id())
     const tx = tokenDAO.transfer(wallet.address, recipient, amount, token) // added token for btc like transfers
     const executeMap = {
       [BLOCKCHAIN_ETHEREUM]: executeTransaction,
       [BLOCKCHAIN_NEM]: executeNemTransaction,
+      [BLOCKCHAIN_BITCOIN]: executeBitccoinTransaction,
```
Comment (wAISw) [inline packages/core/redux/wallets/actions.js]: Add other btc-like blockhains

File: packages/core/redux/bitcoin/constants.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Copyright 2017–2018, LaborX PTY
+ * Licensed under the AGPL Version 3 license.
+ */
+
+export const DUCK_BITCOIN = 'bitcoin'
+
+export const TX_UPDATE = 'tx/update'
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/constants.js]: Let's rename it to BITCOIN_TX_UPDATE and change its value to 'BITCOIN/TX/UPDATE'

File: packages/core/redux/bitcoin/thunks.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  const entry = BitcoinUtils.createBitcoinTxEntryModel({
+    tx: prepared,
+    blockchain: token.blockchain(),
+  }, options)
+
+  dispatch(BitcoinActions.createTransaction(entry))
+  dispatch(submitTransaction(entry))
+}
+
+export const prepareTransaction = (tx, { feeMultiplier = 1, satPerByte = null, symbol }) => async (dispatch, getState) => {
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/thunks.js]: This is not an action. Please move this code into utils.js and call it at line 24 as
`const prepared = BitcoinUtils.prepareTransaction(token, tx, options))`

File: packages/core/redux/bitcoin/thunks.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    to: tx.to,
+    amount: new BigNumber(tx.value),
+    fee: new BigNumber(prepared.fee),
+    prepared: prepared.tx,
+    inputs: prepared.inputs,
+    outputs: prepared.outputs,
+  })
+}
+
+const submitTransaction = (entry) => async (dispatch, getState) => {
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/thunks.js]: This method dispatches only action to open Modal. It is better to move it into src/redux/ui/thunks with appropriate arguments

File: packages/core/redux/bitcoin/thunks.js [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+export const executeBitccoinTransaction = ({ tx, options = null }) => async (dispatch, getState) => {
+  const token = getToken(options.symbol)(getState())
+  const prepared = await dispatch(prepareTransaction(tx, options))
+  const entry = BitcoinUtils.createBitcoinTxEntryModel({
+    tx: prepared,
+    blockchain: token.blockchain(),
+  }, options)
+
+  dispatch(BitcoinActions.createTransaction(entry))
+  dispatch(submitTransaction(entry))
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/thunks.js]: Such methods must be relocated into src/redux/ui. 
```
import * as BitcoinThunks } from '@core'

export const executeBitccoinTransaction = (tx) => (dispatch) => {
  dispatch(BitcoinThunks.executeBitcoinTransaction(tx))
  dispatch(modalsOpen({
    componentName: 'ConfirmTxDialog',
    props: {
      entry,
      description,
      accept: acceptTransaction,
      reject: rejectTransaction,
    },
  }))
}

```

File: packages/core/redux/bitcoin/thunks.js [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+} from '../../models'
+import * as BitcoinActions from './actions'
+import * as BitcoinUtils from './utils'
+import { getSelectedNetwork, getSigner } from '../persistAccount/selectors'
+import { describePendingBitcoinTx } from '../../describers'
+import { getToken } from '../tokens/selectors'
+import { pendingEntrySelector } from './selectors'
+import { notify } from '../notifier/actions'
+
+export const executeBitccoinTransaction = ({ tx, options = null }) => async (dispatch, getState) => {
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/thunks.js]: Typo (double c): executeBitccoinTransaction -> executeBitcoinTransaction

File: packages/core/redux/bitcoin/thunks.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+  }
+}
+
+const sendSignedTransaction = ({ entry }) => async (dispatch, getState) => {
+  dispatch(BitcoinActions.bitcoinTxUpdate(BitcoinUtils.createBitcoinTxEntryModel({
+    ...entry,
+    isPending: true,
+  })))
+
+  // eslint-disable-next-line
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/thunks.js]: It is better to create new variable here. Reassignment does not needed here.

File: packages/core/redux/bitcoin/thunks.js [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+  const node = BitcoinUtils.getNodeByBlockchain(entry.blockchain)
+  if (!node) {
+    dispatch(BitcoinActions.bitcoinTxUpdate(BitcoinUtils.createBitcoinTxEntryModel({
+      ...entry,
+      isErrored: true,
+    })))
+    return null
+  }
+
+  const res = await node.send(entry.tx.from, entry.tx.signed.toHex())
```
Comment (ozalexo) [inline packages/core/redux/bitcoin/thunks.js]: Concept of 'nodes' is deprecated. Please create new module packages/core/redux/bitcoin/http.js
as it was done for profile 

Http requests module (axios): https://github.com/ChronoBank/ChronoMint/blob/develop/packages/core/redux/profile/service.js

Usage example:
https://github.com/ChronoBank/ChronoMint/blob/develop/packages/core/redux/profile/thunks.js

**Comment by ozalexo** [comment]
These changes have been merged into https://github.com/ChronoBank/ChronoMint/pull/787
Closing.

**Pull request closed without merging**